File: codes/Logistic regression/graph_linear_vs_logistic_reg.R
library(tidyverse)

set.seed(123)

x <- rnorm(100)
y <- ifelse(x >0 , 1, 0)

outliers_0 <- ifelse(sample(y, 20) == 0, 1,1)
outliers_1 <- ifelse(sample(y, 20) == 1, 0,0)

x <- c(x,
  rnorm(length(outliers_0)),
  rnorm(length(outliers_1))
  )

y <- c(y,
       outliers_0,
       outliers_1)

plot(x,y)

df <- data.frame(x,y)

###########################################
# linear vs logisitc regression 
###########################################

linear_reg <- ggplot(df, aes(x=x, y=y)) + 
    geom_point(colour = "blue") + 
    stat_smooth(method="glm", se=FALSE,colour = "red")+
    scale_y_continuous(limits=c(-0.2, 1.2),
                       breaks = c(0, 1), 
                       labels = c("0", "1")) +
    theme_classic()

logistic_reg <- ggplot(df, aes(x=x, y=y)) + 
    geom_point(colour = "blue") + 
    stat_smooth(method="glm", method.args = list(family = "binomial"), se=FALSE,colour = "red")+
    scale_y_continuous(limits=c(-0.2, 1.2),
                       breaks = c(0, 1), 
                       labels = c("0", "1")) +
    theme_classic()


sc_plots = list()

sc_plots$linear_reg = linear_reg
sc_plots$logistic_reg = logistic_reg

grid.arrange(sc_plots$linear_reg, sc_plots$logistic_reg, ncol = 2)

###########################################
# logistic regression prediction
###########################################

ggplot(df, aes(x=x, y=y)) + 
    geom_point(colour = "blue") + 
    stat_smooth(method="glm", se=FALSE,colour = "red", size = 1.1)+
    scale_y_continuous(limits=c(-0.2, 1.2),
                       breaks = c(0, 1), 
                       labels = c("Negative Infinity", "Positive Infinity"))  + 
    labs(x = "X variable",
         y = "Log (Odds)") +
    theme_classic()

File: codes/Logistic regression/graph_linear_vs_logistic_reg.py
import pandas as pd
import numpy as np
from random import sample
from plotnine import *
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

n = 1000

x = np.random.normal(0,1,n)

y = [1 if i > 0 else 0 for i in x ]

outliers_0 = [1 if i == 0 else 1 for i in sample(y, int(n*0.05))]

outliers_1 = [0 if i == 1 else 0 for i in sample(y, int(n*0.05))]

x = np.append(x, np.random.normal(0,1,2*len(outliers_0)))
y = y + outliers_0 + outliers_1

# instantiate the model (using the default parameters)
logreg = LogisticRegression()

# fit the model with data
logreg.fit(x.reshape(-1,1),y)

# predictions
preds=logreg.predict_proba(x.reshape(-1,1))

df = pd.DataFrame(dict(y = y, x = x, preds = preds[:,1]))

df.head(10)
###########################################
# linear vs logisitc regression 
###########################################

ggplot(
    data = df,
    mapping = aes(x = 'x', y = 'y')
) + geom_point(colour = "blue") + \
    stat_smooth(method="glm", se=False,colour = "red", size = 0.5) + \
    scale_y_continuous(limits=[-0.2, 1.2],
                       breaks = [0, 1], 
                       labels = ["0", "1"]) + \
    theme_classic() + \
    labs(
        x = "X",
        y = "Y"
    )

    
ggplot(
    data = df,
    mapping = aes(x = 'x', y = 'y')
) + geom_point(colour = "blue") + \
    geom_point(aes(y = 'preds'), colour = 'red', size = 0.5)  + \
    scale_y_continuous(limits=[-0.2, 1.2],
                       breaks = [0, 1], 
                       labels = ["0", "1"]) + \
    theme_classic() + \
    labs(
        x = "X",
        y = "Y"
    )
    
